### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by tpazderka**
- [ ] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [x] The documentation has been updated, if necessary.
- [x] New code is annotated.
- [ ] Changes are covered by tests.
---
Basic idea of settings. Mostly as a proof of concept and for design review.

This attempts to fix several issues regarding documentation (autodoc), too complex init methods and inconsistent settings..

The classes are not complete (they miss any sort of loading from files).

Also the conversion is not finished.

### CODE REVIEW

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- src/oic/utils/settings.py  1
         

Clones added
============
- src/oic/oauth2/base.py  1
- src/oic/oauth2/__init__.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/pyoidc/pyoidc/pullRequest?prid=4592649&bid=15420270)

File: src/oic/utils/settings.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+class PyoidcSettings:
+    """
+    Main class for all settings shared among consumer and client.
+
+    Keyword Args:
+        verify_ssl
+            Control TLS server certificate validation.
+            If set to True the certificate is validated against the global settings,
+            if set to False, no validation is performed.
+            If set to a filename his is used as a certificate bundle in openssl format.
```
Comment (schlenk) [inline src/oic/utils/settings.py]: his => this

**Comment by tpazderka** [comment]
Thanks for the feedback. Will work on this a little bit more to propagate the settings to all necessary classes. We can then expand their content.

**Comment by tpazderka** [comment]
Thanks, will work on this more during the week to get the basic structure.

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oauth2/base.py  1
- src/oic/oauth2/__init__.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/pyoidc/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oauth2/base.py  1
- src/oic/oauth2/__init__.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/pyoidc/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 27
           

Complexity increasing per file
==============================
- src/oic/oic/claims_provider.py  1
- src/oic/oauth2/consumer.py  1
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oauth2/provider.py  2
- src/oic/oic/provider.py  2
- src/oic/oic/__init__.py  4
- src/oic/oauth2/base.py  2
- src/oic/oauth2/__init__.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/tpazderka/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 27
           

Complexity increasing per file
==============================
- src/oic/oauth2/consumer.py  1
- src/oic/oic/claims_provider.py  1
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oauth2/base.py  2
- src/oic/oic/provider.py  2
- src/oic/oauth2/__init__.py  5
- src/oic/oauth2/provider.py  2
- src/oic/oic/__init__.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/tpazderka/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
I think that the basic design of code is done. I will investigate the failure of integration tests, but the code/documentation is probably ready for review.

I am not sure about the type checking in `__setitem__`. If we drop 3.5, we could use [pydantic](https://pydantic-docs.helpmanual.io/) for the same effect. But 3.5 still has support... :(

I would like to add some loading methods and other settings in separate issues.

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 27
           

Complexity increasing per file
==============================
- src/oic/oauth2/consumer.py  1
- src/oic/utils/settings.py  5
- src/oic/oic/claims_provider.py  1
         

Clones added
============
- src/oic/oic/__init__.py  4
- src/oic/oauth2/__init__.py  5
- src/oic/oauth2/provider.py  2
- src/oic/oauth2/base.py  2
- src/oic/oic/provider.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/tpazderka/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 27
           

Complexity increasing per file
==============================
- src/oic/oauth2/consumer.py  1
- src/oic/oic/claims_provider.py  1
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oic/provider.py  2
- src/oic/oic/__init__.py  4
- src/oic/oauth2/provider.py  2
- src/oic/oauth2/base.py  2
- src/oic/oauth2/__init__.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/tpazderka/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Comment by tpazderka** [comment]
OK, test are fixed. I am going to ignore codacy errors as they are mostly false positives.

**Comment by tpazderka** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/oic/oic/claims_provider.py  1
- src/oic/oauth2/consumer.py  1
- src/oic/utils/settings.py  5
         

Clones added
============
- src/oic/oauth2/base.py  2
- src/oic/oauth2/__init__.py  5
- src/oic/oic/__init__.py  4
- src/oic/oic/provider.py  2
- src/oic/oauth2/provider.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/tpazderka/pyoidc/pullRequest?prid=4592649&bid=15420270)

**Pull request merged**